### PR TITLE
feature/improve binder

### PR DESCRIPTION
**Pull request by HaronCode**
* add new binder implementation 
* global refactoring 
* improve all tests and add new
* update view model module

### CODE REVIEW

File: gemini-core/src/test/kotlin/com/haroncode/gemini/common/TestConnectionBinder.kt [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+    fun disconnectAll() = testLifecycle.disconnectAll()
+
+    override fun bind(connectionRule: ConnectionRule) = binder.bind(connectionRule)
+
+    override fun isDisposed() = binder.isDisposed
+
+    override fun dispose() = binder.dispose()
+
+    private class TestLifeCycle : StoreLifecycle {
```
Comment (kdk96) [inline gemini-core/src/test/kotlin/com/haroncode/gemini/common/TestConnectionBinder.kt]: LifeCycle -> Lifecycle

File: gemini-core/src/test/kotlin/com/haroncode/gemini/core/BaseStoreTest.kt [hunk 2]
```diff
@@ -21,139 +20,199 @@ import java.util.concurrent.TimeUnit
...
     @Test
     fun `emitted initial state is correct`() {
-        val state = states.onNextEvents().first()
+        val state = testStatesObserver.onNextEvents().first()
         assertEquals(INITIAL_COUNTER, state.counter)
         assertEquals(INITIAL_LOADING, state.loading)
     }
 
+    @Test
+    fun `bootstraper correct work after create store`() {
```
Comment (kdk96) [inline gemini-core/src/test/kotlin/com/haroncode/gemini/core/BaseStoreTest.kt]: bootstrapper

File: gemini-core/src/test/kotlin/com/haroncode/gemini/core/BaseStoreTest.kt [hunk 3]
```diff
@@ -21,139 +20,199 @@ import java.util.concurrent.TimeUnit
...
+            FulfillableInstantly
+        )
+
+        actions.forEach(testBootstrapperSubject::onNext)
+
+        assertEquals(2, testStatesObserver.onNextEvents().size)
+    }
+
+    @Test
+    fun `event producer doesn't react on all effect`() {
```
Comment (kdk96) [inline gemini-core/src/test/kotlin/com/haroncode/gemini/core/BaseStoreTest.kt]: all effects

File: gemini-core/src/test/kotlin/com/haroncode/gemini/core/BaseStoreTest.kt [hunk 4]
```diff
@@ -21,139 +20,199 @@ import java.util.concurrent.TimeUnit
...
+        )
+
+        val events = Flowable.fromPublisher(baseStore.eventSource).test()
+        actions.forEach(testActionsSubject::onNext)
+
+        events.assertNoValues()
+    }
+
+    @Test
+    fun `event producer react on special effect`() {
```
Comment (kdk96) [inline gemini-core/src/test/kotlin/com/haroncode/gemini/core/BaseStoreTest.kt]: reacts

File: gemini-android/build.gradle [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+            proguardFiles getDefaultProguardFile('proguard-android.txt'), 'proguard-rules.pro'
+        }
+    }
+}
+
+dependencies {
+    compileOnly rootProject.ext.appCompat
+    implementation rootProject.ext.kotlinStdlib
+    /**
+     * Rx java
```
[the remark this answers is not in the document]
Comment (HaronCode) [reply gemini-android/build.gradle]: done

File: gemini-core/src/test/kotlin/com/haroncode/gemini/binder/BaseConnectionBinderTest.kt [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+    @Before
+    fun setUp() {
+        testStoreLifecycleProcessor = PublishProcessor.create()
+        testStoreLifecycle = TestStoreLifecycle(testStoreLifecycleProcessor)
+
+        testConnectionBinder = BaseConnectionBinder(testStoreLifecycle)
+    }
+
+    @Test
+    fun `binder subscribe if lifecycle is active`() {
```
[the remark this answers is not in the document]
Comment (HaronCode) [reply gemini-core/src/test/kotlin/com/haroncode/gemini/binder/BaseConnectionBinderTest.kt]: done

File: gemini-core/src/test/kotlin/com/haroncode/gemini/binder/BaseConnectionBinderTest.kt [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+
+        testStoreLifecycleProcessor.onNext(Event.START)
+        testConnectionBinder.bind(connectionRule)
+        testStoreLifecycleProcessor.onNext(Event.STOP)
+
+        Assert.assertTrue(connectionRule.isConnected.get())
+    }
+
+    @Test
+    fun `binder dispose all connection if event stop`() {
```
[the remark this answers is not in the document]
Comment (HaronCode) [reply gemini-core/src/test/kotlin/com/haroncode/gemini/binder/BaseConnectionBinderTest.kt]: done

File: gemini-core/src/test/kotlin/com/haroncode/gemini/binder/BaseConnectionBinderTest.kt [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+        testStoreLifecycleProcessor.onNext(Event.STOP)
+
+        testStoreLifecycleProcessor.onNext(Event.START)
+
+        val allConnected = connections.all { connection -> connection.isConnected.get() }
+        Assert.assertTrue(allConnected)
+    }
+
+    @Test
+    fun `binder distinct events`() {
```
[the remark this answers is not in the document]
Comment (HaronCode) [reply gemini-core/src/test/kotlin/com/haroncode/gemini/binder/BaseConnectionBinderTest.kt]: done

File: gemini-core/src/test/kotlin/com/haroncode/gemini/binder/BaseConnectionBinderTest.kt [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+    }
+
+    @Test
+    fun `binder subscribe if lifecycle is active`() {
+        val connectionRule = TestConnectionRule(false)
+
+        testStoreLifecycleProcessor.onNext(Event.START)
+        testConnectionBinder.bind(connectionRule)
+
+        Assert.assertTrue(connectionRule.isConnected.get())
```
[the remark this answers is not in the document]
Comment (HaronCode) [reply gemini-core/src/test/kotlin/com/haroncode/gemini/binder/BaseConnectionBinderTest.kt]: done

File: gemini-core/src/test/kotlin/com/haroncode/gemini/common/TestConnectionBinder.kt (hunk 1 above)
Comment (HaronCode) [reply gemini-core/src/test/kotlin/com/haroncode/gemini/common/TestConnectionBinder.kt]: done

File: gemini-core/src/test/kotlin/com/haroncode/gemini/core/BaseStoreTest.kt (hunk 2 above)
Comment (HaronCode) [reply gemini-core/src/test/kotlin/com/haroncode/gemini/core/BaseStoreTest.kt]: done

File: gemini-core/src/test/kotlin/com/haroncode/gemini/core/BaseStoreTest.kt (hunk 3 above)
Comment (HaronCode) [reply gemini-core/src/test/kotlin/com/haroncode/gemini/core/BaseStoreTest.kt]: done

File: gemini-core/src/test/kotlin/com/haroncode/gemini/core/BaseStoreTest.kt (hunk 4 above)
Comment (HaronCode) [reply gemini-core/src/test/kotlin/com/haroncode/gemini/core/BaseStoreTest.kt]: done

File: gemini-android/src/main/java/com/haroncode/gemini/android/binder/MainThreadConnectionBinder.kt [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+import android.os.Looper
+import com.haroncode.gemini.core.ConnectionBinder
+import com.haroncode.gemini.core.ConnectionRule
+
+class MainThreadConnectionBinder(
+    private val connectionBinder: ConnectionBinder
+) : ConnectionBinder {
+
+    override fun bind(connectionRule: ConnectionRule) {
+        check(Looper.myLooper() == Looper.getMainLooper()) { "Method bind() can invoked only in main thread" }
```
[the remark this answers is not in the document]
Comment (HaronCode) [reply gemini-android/src/main/java/com/haroncode/gemini/android/binder/MainThreadConnectionBinder.kt]: done

**Pull request merged**